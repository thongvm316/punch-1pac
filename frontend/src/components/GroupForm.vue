<template>
  <div>
    <div
      class="form-group"
      :class="{ 'has-error': $v.params.name.$error || errors.name }"
    >
      <label class="form-label">{{ $t('label.name') }}</label>
      <input
        v-model="$v.params.name.$model"
        class="form-input"
        type="text"
        :placeholder="$t('label.name')"
      >
      <p
        v-if="$v.params.name.$error && !errors.name"
        class="form-input-hint"
      >
        {{ $t('validation.required', { name: $t('label.name') }) }}
      </p>
      <p
        v-if="errors.name"
        class="form-input-hint"
      >
        {{ $t('label.name') }} {{ errors.name[0] }}
      </p>
    </div>
    <div
      class="form-group"
      :class="{ 'has-error': errors.image }"
    >
      <label class="form-label">{{ $t('label.image') }}</label>
      <img
        v-if="targetGroup"
        class="img-profile"
        :src="targetGroup.image_url"
        alt=""
      >
      <input
        class="form-input"
        type="file"
        @change="setImageFile"
      >
      <p
        v-if="errors.image"
        class="form-input-hint"
      >
        {{ $t('label.image') }} {{ errors.image[0] }}
      </p>
    </div>
    <div
      class="form-group"
      :class="{ 'has-error': errors.description }"
    >
      <label class="form-label">{{ $t('label.description') }}</label>
      <textarea
        v-model="params.description"
        class="form-input"
        :placeholder="$t('label.description')"
      />
      <p
        v-if="errors.description"
        class="form-input-hint"
      >
        {{ $t('label.description') }} {{ errors.description[0] }}
      </p>
    </div>
    <div class="form-group">
      <button
        v-if="!targetGroup"
        ref="localAddGroupButton"
        type="button"
        class="btn btn-success btn-submit"
        :disabled="isDisabled"
        @click="localAddGroup"
      >
        {{ $t('button.common.submit') }}
      </button>
      <button
        v-if="targetGroup"
        ref="localEditGroupButton"
        type="button"
        class="btn btn-success btn-submit"
        :disabled="isDisabled"
        @click="localEditGroup"
      >
        {{ $t('button.common.save') }}
      </button>
    </div>
  </div>
</template>

<script>
import { mapActions } from 'vuex'
import handleSuccess from '@/mixins/handle-success'
import groupFormValidate from '@/validations/group-form-validate'

export default {
  name: 'GroupForm',

  mixins: [handleSuccess, groupFormValidate],

  props: {
    targetGroup: {
      type: Object,
      default: null
    }
  },

  data() {
    return {
      params: {
        name: '',
        description: '',
        image: ''
      },
      errors: {}
    }
  },

  created() {
    if (this.targetGroup) this.params = { ...this.targetGroup }
  },

  methods: {
    ...mapActions('groups', ['addGroup']),

    ...mapActions('group', ['updateGroup']),

    localAddGroup() {
      this.addGroup(this.params)
        .then(() => {
          this.handleSuccess({
            emitType: 'afterModify',
            message: this.$t('messages.group.createSuccess')
          })
        })
        .catch(error => {
          if (error.response && error.response.status === 422) this.errors = error.response.data.errors
        })
    },

    localEditGroup() {
      this.updateGroup({ groupId: this.targetGroup.id, editParams: this.params })
        .then(() => {
          this.handleSuccess({
            emitType: 'afterModify',
            message: this.$t('messages.group.updateSuccess')
          })
        })
        .catch(error => {
          if (error.response && error.response.status === 422) this.errors = error.response.data.errors
        })
    },

    setImageFile(e) {
      const files = e.target.files || e.dataTransfer.files
      if (!files.length) return
      this.params.image = files[0]
    }
  }
}
</script>
