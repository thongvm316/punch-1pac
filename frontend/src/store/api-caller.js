import axios from 'axios'

export default function callApi(payload = { params: null, data: null }) {
  const { method, url, params, data } = payload

  return axios({
    method,
    url,
    params,
    data
  })
}
