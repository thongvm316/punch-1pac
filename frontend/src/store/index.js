import Vue from 'vue'
import Vuex from 'vuex'
import createLogger from 'vuex/dist/logger'

import headerAnnouncements from './modules/header-announcements'
import punch from './modules/punch'

import companyUsers from './modules/company-users'
import companyDepartments from './modules/company-departments'

import announcements from './modules/announcements'
import announcement from './modules/announcement'
import userSessions from './modules/user-sessions'
import attendances from './modules/attendances'
import companyAllowedIPs from './modules/company-allowed_ips'

Vue.use(Vuex)

const debug = process.env.NODE_ENV !== 'production'

export default new Vuex.Store({
  modules: {
    headerAnnouncements,
    punch,
    companyUsers,
    companyDepartments,
    announcements,
    announcement,
    userSessions,
    attendances,
    companyAllowedIPs
  },
  strict: debug,
  plugins: debug ? [createLogger()] : []
})
