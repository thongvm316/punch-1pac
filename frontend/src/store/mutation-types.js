// initial-states
export const INITIAL_STATES_SET_USER = 'INITIAL_STATES_SET_USER'
export const INITIAL_STATES_SET_COMPANY = 'INITIAL_STATES_SET_COMPANY'
export const INITIAL_STATES_SET_META = 'INITIAL_STATES_SET_META'
export const INITIAL_STATES_UPDATE_USER = 'INITIAL_STATES_UPDATE_USER'
export const INITIAL_STATES_UPDATE_COMPANY = 'INITIAL_STATES_UPDATE_COMPANY'
export const INITIAL_STATES_SET_USER_ERRORS = 'INITIAL_STATES_SET_USER_ERRORS'
export const INITIAL_STATES_SET_COMPANY_ERRORS = 'INITIAL_STATES_SET_COMPANY_ERRORS'
export const INITIAL_STATES_CLEAR_USER_ERRORS = 'INITIAL_STATES_CLEAR_USER_ERRORS'
export const INITIAL_STATES_CLEAR_COMPANY_ERRORS = 'INITIAL_STATES_CLEAR_COMPANY_ERRORS'
export const INITIAL_STATES_UPDATE_PASSWORD_CHANGED = 'INITIAL_STATES_UPDATE_PASSWORD_CHANGED'

// announcements
export const READ_ANNOUNCEMENT = 'READ_ANNOUNCEMENT'
export const RECEIVE_HEADER_ANNOUNCEMENTS = 'RECEIVE_HEADER_ANNOUNCEMENTS'

// attendances
export const RECEIVE_ATTENDANCES = 'RECEIVE_ATTENDANCES'

// punch
export const PUNCH_IN = 'PUNCH_IN'
export const PUNCH_OUT = 'PUNCH_OUT'
export const PUNCH_INIT_ATTENDANCE = 'PUNCH_INIT_ATTENDANCE'

// company-users
export const FETCH_USERS = 'FETCH_USERS'
export const DELETE_USER = 'DELETE_USER'
export const UPDATE_USER = 'UPDATE_USER'

// company-business-days
export const FETCH_BUSINESS_DAYS = 'FETCH_BUSINESS_DAYS'
export const ADD_BUSINESS_DAY = 'ADD_BUSINESS_DAY'
export const DELETE_BUSINESS_DAY = 'DELETE_BUSINESS_DAY'
export const UPDATE_BUSINESS_DAY = 'UPDATE_BUSINESS_DAY'
export const UPDATE_BUSINESS_DAY_ERRORS = 'UPDATE_BUSINESS_DAY_ERRORS'
export const CLEAR_BUSINESS_DAY_ERRORS = 'CLEAR_BUSINESS_DAY_ERRORS'

// company-allowed_ips
export const FETCH_IPS = 'FETCH_IPS'
export const CREATE_IP = 'CREATE_IPS'
export const UPDATE_IP = 'UPDATE_IPS'
export const DELETE_IP = 'DELETE_IPS'
export const UPDATE_IP_ERRORS = 'UPDATE_REQUEST_ERRORS'
export const CLEAR_IP_ERRORS = 'CLEAR_REQUEST_ERRORS'

// company-holidays
export const FETCH_HOLIDAYS = 'FETCH_HOLIDAYS'
export const CREATE_HOLIDAY = 'CREATE_HOLIDAY'
export const UPDATE_HOLIDAY = 'UPDATE_HOLIDAY'
export const DELETE_HOLIDAY = 'DELETE_HOLIDAY'
export const UPDATE_HOLIDAY_ERRORS = 'UPDATE_HOLIDAY_ERRORS'
export const CLEAR_HOLIDAY_ERRORS = 'CLEAR_HOLIDAY_ERRORS'
export const IMPORT_NATIONAL_HOLIDAYS = 'IMPORT_NATIONAL_HOLIDAYS'

// user-sessions
export const FETCH_SESSIONS = 'FETCH_SESSIONS'
export const DELETE_SESSION = 'DELETE_SESSION'

// user-password
export const UPDATE_USER_PASSWORD_ERRORS = 'UPDATE_USER_PASSWORD_ERRORS'
export const CLEAR_USER_PASSWORD_ERRORS = 'CLEAR_USER_PASSWORD_ERRORS'

// requests
export const RECEIVE_REQUESTS = 'RECEIVE_REQUESTS'
export const ADD_REQUEST = 'ADD_REQUEST'
export const UPDATE_REQUEST = 'UPDATE_REQUEST'
export const DELETE_REQUEST = 'DELETE_REQUEST'
export const UPDATE_REQUEST_ERRORS = 'UPDATE_REQUEST_ERRORS'
export const CLEAR_REQUEST_ERRORS = 'CLEAR_REQUEST_ERRORS'

// group-requests
export const RECEIVE_GROUP_REQUESTS = 'RECEIVE_GROUP_REQUESTS'
export const APPROVE_GROUP_REQUEST = 'APPROVE_GROUP_REQUEST'
export const REJECT_GROUP_REQUEST = 'REJECT_GROUP_REQUEST'
export const REJECT_GROUP_REQUEST_ERRORS = 'REJECT_GROUP_REQUEST_ERRORS'
export const CLEAR_REJECT_GROUP_REQUEST_ERRORS = 'CLEAR_REJECT_GROUP_REQUEST_ERRORS'

// group-attendances
export const RECEIVE_GROUP_ATTENDANCES = 'RECEIVE_GROUP_ATTENDANCES'
export const FETCH_USERS_IN_GROUP_ATTENDANCES = 'FETCH_USERS_IN_GROUP_ATTENDANCES'

// group-report
export const FETCH_GROUP_REPORT = 'FETCH_GROUP_REPORT'

// groups
export const RECEIVE_GROUPS = 'RECEIVE_GROUPS'
export const ADD_GROUP = 'ADD_GROUP'
export const UPDATE_GROUPS_ERRORS = 'UPDATE_GROUPS_ERRORS'
export const CLEAR_GROUPS_ERRORS = 'CLEAR_GROUPS_ERRORS'

// group
export const RECEIVE_GROUP = 'RECEIVE_GROUP'
export const UPDATE_GROUP = 'UPDATE_GROUP'
export const ADD_GROUP_USER = 'ADD_GROUP_USER'
export const REMOVE_GROUP_USER = 'REMOVE_GROUP_USER'
export const UPDATE_GROUP_ERRORS = 'UPDATE_GROUP_ERRORS'
export const CLEAR_GROUP_ERRORS = 'CLEAR_GROUP_ERRORS'
export const FETCH_USERS_NOT_IN_GROUP = 'FETCH_USERS_NOT_IN_GROUP'

// chart
export const FETCH_CHART_DATA = 'FETCH_CHART_DATA'
export const RESET_CHART_DATA = 'RESET_CHART_DATA'

// calendar
export const FETCH_CALENDAR_ATTENDANCES = 'FETCH_CALENDAR_ATTENDANCES'

// flash
export const SET_FLASH_MESSAGE = 'SET_FLASH_MESSAGE'

// notifications
export const FETCH_HEADER_NOTIFICATIONS = 'FETCH_HEADER_NOTIFICATIONS'
export const FETCH_NEW_HEADER_NOTIFICATIONS = 'FETCH_NEW_HEADER_NOTIFICATIONS'
export const FETCH_MORE_HEADER_NOTIFICATIONS = 'FETCH_MORE_HEADER_NOTIFICATIONS'
export const READ_NOTIFICATIONS = 'READ_NOTIFICATIONS'
export const SET_INTERVAL_FETCH_NOTIFICATIONS = 'SET_INTERVAL_FETCH_NOTIFICATIONS'

// activities
export const FETCH_ACTIVITIES = 'FETCH_ACTIVITIES'
export const FETCH_MORE_ACTIVITIES = 'FETCH_MORE_ACTIVITIES'

// status-cards
export const FETCH_STATUS_CARDS = 'FETCH_STATUS_CARDS'
