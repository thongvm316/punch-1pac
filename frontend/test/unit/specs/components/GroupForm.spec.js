import { shallowMount } from '@vue/test-utils'

import localVue from '../../supports/local-vue'
import setComputed from '../../supports/set-computed'

import store from '@/store'
import i18n from '@/locale'
import GroupForm from '@/components/GroupForm'

const targetGroup = {
  name: '1pacvn',
  description: '1pacvn team',
  image: '/'
}

const localAddGroup = jest.fn()
const localEditGroup = jest.fn()

describe('GroupForm.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(GroupForm, {
      i18n,
      store,
      methods: { localAddGroup },
      localVue
    })
  })

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when GroupForm was mounted', () => {
    it('should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.isVueInstance()).toBeTruthy()
    })

    it('should render created GroupForm', () => {
      expect(wrapper.findAll('.form-group')).toHaveLength(4)
      expect(wrapper.findAll('.form-group button')).toHaveLength(1)
      expect(wrapper.find({ ref: 'localAddGroupButton' }).exists()).toBe(true)
    })
  })

  describe('when errors', () => {
    it('should render no form-input-hint error text', () => {
      expect(wrapper.findAll('p.form-input-hint')).toHaveLength(0)
    })

    it('should render errors text', async () => {
      const errors = {
        name: ['have been taken'],
        image: ['too large'],
        description: ['wrong']
      }
      wrapper.setData({ errors })
      await wrapper.vm.$nextTick()

      expect(wrapper.findAll('p.form-input-hint')).toHaveLength(3)
      expect(wrapper.findAll('p.form-input-hint').at(0).text()).toEqual('Name have been taken')
      expect(wrapper.findAll('p.form-input-hint').at(1).text()).toEqual('Image too large')
      expect(wrapper.findAll('p.form-input-hint').at(2).text()).toEqual('Description wrong')
    })

    it('should render 1 error', async () => {
      const errors = {
        name: ['have been taken']
      }
      wrapper.setData({ errors })
      await wrapper.vm.$nextTick()

      expect(wrapper.findAll('p.form-input-hint')).toHaveLength(1)
      expect(wrapper.findAll('p.form-input-hint').at(0).text()).toEqual('Name have been taken')
    })
  })

  describe('when addGroup method', () => {
    it('should call localAddGroup methods', () => {
      wrapper.find({ ref: 'localAddGroupButton' }).trigger('click')
      expect(localAddGroup).toHaveBeenCalled()
    })
  })
})

describe('when GroupForm have props data', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(GroupForm, {
      i18n,
      store,
      methods: { localEditGroup },
      propsData: { targetGroup },
      localVue
    })
  })

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when render GroupForm', () => {
    it('should render edit Group Form', () => {
      expect(wrapper.vm.params).toEqual(targetGroup)
      expect(wrapper.findAll('.form-group')).toHaveLength(4)
      expect(wrapper.findAll('.form-group button')).toHaveLength(1)
      expect(wrapper.find({ ref: 'localEditGroupButton' }).exists()).toBe(true)
    })
  })

  describe('when editGroup method', () => {
    it('should call localEditGroup methods', () => {
      wrapper.find({ ref: 'localEditGroupButton' }).trigger('click')
      expect(localEditGroup).toHaveBeenCalled()
    })
  })
})
