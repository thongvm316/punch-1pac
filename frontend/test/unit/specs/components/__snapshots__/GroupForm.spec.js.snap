// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupForm.vue when GroupForm was mounted should render correctly 1`] = `
<div>
  <div class="form-group"><label class="form-label">Name</label> <input type="text" placeholder="Name" class="form-input">
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Image</label>
    <!----> <input type="file" class="form-input">
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Description</label> <textarea placeholder="Description" class="form-input"></textarea>
    <!---->
  </div>
  <div class="form-group"><button type="button" class="btn btn-success btn-submit">
      Submit
    </button>
    <!---->
  </div>
</div>
`;

exports[`GroupForm.vue when GroupForm was mounted should render exists groups 1`] = `
<div>
  <div class="form-group"><label class="form-label">Name</label> <input type="text" placeholder="Name" class="form-input">
    <!---->
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Image</label> <img alt="" class="img-profile"> <input type="file" class="form-input">
    <!---->
  </div>
  <div class="form-group"><label class="form-label">Description</label> <textarea placeholder="Description" class="form-input"></textarea>
    <!---->
  </div>
  <div class="form-group">
    <!----> <button type="button" class="btn btn-success btn-submit">
      Save
    </button></div>
</div>
`;

exports[`GroupForm.vue when errors should render errors text 1`] = `
<div>
  <div class="form-group has-error"><label class="form-label">Name</label> <input type="text" placeholder="Name" class="form-input">
    <!---->
    <p class="form-input-hint">
      Name have been taken
    </p>
  </div>
  <div class="form-group has-error"><label class="form-label">Image</label>
    <!----> <input type="file" class="form-input">
    <p class="form-input-hint">
      Image too large
    </p>
  </div>
  <div class="form-group has-error"><label class="form-label">Description</label> <textarea placeholder="Description" class="form-input"></textarea>
    <p class="form-input-hint">
      Description wrong
    </p>
  </div>
  <div class="form-group"><button type="button" class="btn btn-success btn-submit">
      Submit
    </button>
    <!---->
  </div>
</div>
`;
